### ISSUE TITLE
KDC: TGS-REQ problem with Windows 11 22H2 client

### ISSUE THREAD

**Issue by branciar** [heimdal/heimdal#1011]
We have an Active Directory domain with a outgoing trust relationship to a Heimdal KDC Kerberos realm.

Users authenticate on the Heimdal realm from AD-joined workstations.

We just encountered a problem with the latest Windows 11 version (22H2):

the TGS-REQ issued by the Windows client to the Heimdal KDC for the krbtgt/AD-DOMAIN@HEIMDAL-DOMAIN principal fails with "Decrypt integrity check failed":

Sep 22 12:24:42 mykdc kdc[683]: Failed to verify authenticator checksum: Decrypt integrity check failed for checksum type rsa-md5, key type aes256-cts-hmac-sha1-96
Sep 22 12:24:42 mykdc kdc[683]: Failed parsing TGS-REQ from IPv6:xx:xx:xx:xx:xx:xx:xx:xx
Sep 22 12:24:42 mykdc kdc[683]: tgs-req: sending error: -1765328353 to client

Under exactly same conditions, previous versions of Windows (up to Win11 21H2) obtain a successful TGS-REP from KDC.

Thorough wireshark inspection of the TGS-REQ packets from various Windows clients highlights only one difference :

the "till" timestamp of the "req-body" section is set to 99990913024805Z (9999-09-13 02:48:05 UTC) by the 22H2 Win11 client, whereas older Windows clients set it to 20370913024805Z (2037-09-13 02:48:05 UTC).

This let me suspect a trivial 32-bit Unix epoch counter overflow in KDC...

If so, this bug only occurs on TGS-REQ: in the logon process, Windows first sends AS-REQ containing the same odd timestamp (99990913024805Z value in the "till" and "rtime" fields of the req-body), which succeeds anyway.

Our KDC version is 7.7.0, on x64 Debian 11 platform (heimdal-kdc 7.7.0+dfsg-2 Debian package).

**Comment by jaltman** [comment]
Heimdal attempted to increase the default requested maximum ticket lifetime and maximum renew lifetime to 0x7FFFFFFF seconds in a013e93e95a826b321870a7e626e2cdc9be13984.   This change had to be reverted because it broke the ability to request tickets.  See issue https://github.com/heimdal/heimdal/issues/220.  

A case should be opened with Microsoft to revert their change.   Requesting a maximum ticket lifetime greater than 9999 years is absolutely nuts.    
  
    9999*365.25*24*60*60 = 315544442400 = 0x49:77EA1620
    
and is utterly incompatible with builds for systems that use a signed 32-bit time_t value.  

There is no justification for requesting such a long ticket lifetime in Kerberos.   

Heimdal is not going to be able to fix this in the Heimdal 7.x series.

**Comment by lhoward** [comment]
The reporter is using Heimdal on a 64-bit platform though, so could this be a separate issue?

**Comment by jaltman** [comment]
> The reporter is using Heimdal on a 64-bit platform though, so could this be a separate issue?

From the perspective of Kerberos clients interacting with deployed infrastructure, it is not safe to request tickets with such a large lifetime.  Therefore, Microsoft must roll back the change.

**Comment by jaltman** [comment]
20370913024805Z (2037-09-13 02:48:05 UTC) is 2136422885 = 0x7F5739E5 seconds since Unix Epoch.   It slightly more than 128 days before the 2038 roll over.   I strongly suspect that Microsoft previously knew there was an interop problem and that knowledge was lost.

**Comment by jaltman** [comment]
Is this problem introduced by the Windows 11 22H2 feature update or one of the monthly cumulative updates?

**Comment by jaltman** [comment]
A contact within Microsoft reports that 22H2 removed a cap on the maximum timestamp that would be encoded using ASN.1.   The change is permitted by protocol and therefore would require customers to open an escalation case in order to revert the change now that it has shipped to customers.

**Comment by branciar** [comment]
Requesting tickets with "unlimited" lifetime is not so uncommon. 
Heimdal's own "kgetcred" and Samba's "smbclient" commands both issue TGS-REQ with a "till" set to 19700101000000Z, which is assumed as MAX_TIME by _kdc_fix_time() (https://github.com/heimdal/heimdal/blob/master/kdc/kerberos5.c#L57-L64).
Indeed, RFC 4120 definition of "till" says : "... if the requested endtime is "19700101000000Z", the requested ticket is to have the maximum endtime permitted according to KDC policy."
Microsoft's fault only appears to have chosen an arbitrary large "till" to request "maximum endtime permitted according to KDC policy", instead of using this special-purpose "epoch-style" value.
Anyway, it's the KDC job to cap the requested ticket lifetime according to it's own policy.

**Comment by prigaux** [comment]
Looking at the code, the issue may comes from https://github.com/heimdal/heimdal/blob/master/lib/asn1/timegm.c :

>      Refuse to calculate time ~ 2000 years into the future, this is
>      not possible for systems where time_t is a int32_t, however,
>      when time_t is a int64_t, that can happen, and this becomes a
>      denial of sevice.

tm_year = 9999 - 1900 = 8099
is compared to
ASN1_MAX_YEAR = 2000

Hence the _der_timegm returns 0.

(I do not understand why `_kdc_fix_time` does not set it to `MAX_TIME` though...)

**Comment by jaltman** [comment]
> Heimdal's own "kgetcred" and Samba's "smbclient" commands both issue TGS-REQ with a "till" set to 19700101000000Z, which is assumed as MAX_TIME by _kdc_fix_time() (https://github.com/heimdal/heimdal/blob/master/kdc/kerberos5.c#L57-L64).

19700101000000Z is UNIX Epoch which is 0 seconds.

> In `_der_timegm` : tm_year = 9999 - 1900 = 8099 is compared to ASN1_MAX_YEAR = 2000
> 
> Hence `_der_timegm` returns 0.

which is UNIX Epoch.   This should be the same value that would be the result of receiving 19700101000000Z.
 
> (I do not understand why `_kdc_fix_time` does not set it to `MAX_TIME` though...)

The error is being generated as a result of `_kdc_verify_checksum()` failing when called from `tgs_check_authenticator()` which is called from `tgs_parse_request()`.   The `_kdc_fix_time()` calls wouldn't take place until `tgs_make_reply()`.   The failure occurs before the conversion from UNIX Epoch to MAX_TIME would take place.  

The Heimdal ASN.1 engine does not support timestamps larger than 0x7FFFFFFF seconds from UNIX Epoch.  I suspect what is happening is that the ASN.1 conversion of 0x7F5739E5 seconds to 0x0 seconds for both the ticket endtime and renewtime is causing the checksum validation to fail.

Modifying the ASN.1 engine and the internals of libkrb5 and the kdc to support 64-bit time is a non-trivial amount of work.   https://github.com/heimdal/heimdal/pull/743 is the in-progress effort but it is far from complete.

**Comment by jaltman** [comment]
I have filed a report with the Microsoft Security Response Center.   Microsoft customer reports are encouraged.

**Comment by SteveSyfuhs** [comment]
Howdy folks. One of the Windows Kerberos owners here. Couple points.

1. This is not a security issue. Moreso this is not a security issue in any of _our_ products. MSRC will close it as by design. Just an FYI for any future communications (they have no authority to impose change on things that are not security-related).
2. This is not our problem. Our KDC, our client, and a handful of other well-known implementations have no issue with this change.
3. This is not out of spec. RFC pretty clearly skips any mention of maximum values, and any meaningful interpretation says it should just follow the `GeneralizedTime` ranges.
4. Everyone has 16 years to fix the 2038 bug and we know how long it takes the industry to move off unsupported or unpatched versions of software. Better to nip this early.

We can fix this, but we are reluctant to do anything that requires us to limit any value to 0xffffffff.

Also, any 2038 logic we had in the past has always been to handle interop scenarios where implementations use 32 bit values for timing. We have never had this problem in Windows because all of our time is 64 bit in NTOS.

**Comment by greghudson** [comment]
> This is not our problem. Our KDC, our client, and a handful of other well-known implementations have no issue with this change.

I'm not sure if "a handful of other well-known implementations" refers to MIT krb5 here.  MIT krb5 can represent timestamps up to 21060207062815Z, as of release 1.16 (prior to that it is limited to y2038).  I tested what happens if a client sends 99990913024805Z to the current MIT krb5 KDC in an AS-REQ or TGS-REQ, and it appears to work, but I'm not sure I would say that it has "no issue".  The timestamp is truncated to 32 bits and is therefore interpreted as 21051011112837Z.  Fortunately the current KDC code does not appear to re-encode the KDC-REQ-BODY for any checksums.

**Comment by branciar** [comment]
According to our investigations, the "Failed to verify authenticator checksum" bug is directly related to the ASN1_MAX_YEAR limitation in _der_gmtime from lib/asn1/timegm.c.

To verify checksum, the parsed req-body is first encoded using ASN1_MALLOC_ENCODE, which indirectly calls _heim_time2generalizedtime which depends on  _der_gmtime. 
If the "till" timestamp is out of _der_gmtime bounds, the encoded req-body is altered, and krb5_verify_checksum fails.

As a proof of concept, we replaced the bodies of _der_gmtime and _der_timegm by direct calls to standard time.h  gmtime_r and timegm in lib/asn1/timegm.c:

```
#include "der_locl.h"
#include <time.h>

time_t
_der_timegm (struct tm *tm)
{
    if (tm->tm_mday < 1) return -1; /* to get check-timegm.c happy */
    return timegm(tm);
}

struct tm *
_der_gmtime(time_t t, struct tm *tm)
{
    return gmtime_r(&t,tm);
}
```

This resolves the bug, now TGS-REQ using till=99990913024805Z are honored and don't fail checksum anymore.

This is a quick hack. 

As a better approach, given that _der_gmtime and _der_timegm are only used by _heim_time2generalizedtime (lib/asn1/der_put.c)  and generalizedtime2time (lib/asn1/der_get.c), it would be possible to :
- discard lib/asn1/timegm.c
- modify _heim_time2generalizedtime and generalizedtime2time to directly call time.h functions (timegm, gmtime_r)
- and implement ASN.1 timestamp boundary checks directly in those functions. This would be far better than the rough checks done by timegm.c, since _heim_time2generalizedtime and generalizedtime2time are aware of the timestamp variant handled (generalizedTime, or UTCTime), which have quite different boundaries.
- and of course, get rid of the check-timegm.c. Maybe it could be replaced by a more general time2generalizedtime check.

Another option would be to run krb5_verify_checksum directly on the incoming binary req-body, and note a re-encoded decoded one. But this involves quite more code rework. And anyway, ASN.1 timestamp boundary checks need to be updated.

**Comment by jaltman** [comment]
@branciar, Thank you for your continued investigations.   I believe that one of the reasons that Heimdal does not use OS timegm() and gmtime() in ASN.1 is avoiding incorporation of leap seconds in KerberosTime.   Heimdal has used OS versions of timegm/gmtime elsewhere and has implemented compatibility implementations in libroken.  However, it has explicitly not used them in the ASN.1 library for KerberosTime.  Unfortunately, the relevant RFCs are ambiguous on the subject.   MIT krb5 also used its own timegm|gmtime equivalents for ASN.1 up until 2011 or 2012.   In one of the associated commit messages tlyu@mit.edu mentioned the leap seconds issue and points out that most problems associated with the mixture of leap seconds and non-leap seconds would likely be masked by clock skew tolerances.  

Even though the Heimdal master branch has the same signed 32-bit time issues as the Heimdal heimdal-7-1-branch we do not believe it experiences the observed checksum failure because the authenticator contents are not re-encoded on the master branch as they are on the heimdal-7-1-branch.   

@branciar wrote:

> Another option would be to run krb5_verify_checksum directly on the incoming binary req-body, and not a decoded/re-encoded one. But this involves quite more code rework. And anyway, ASN.1 timestamp boundary checks need to be updated.

This is in fact the correct solution and is the behavior that is present on the Heimdal master branch.   There is not a clean cherry-pick of the changes to the heimdal-7-1-branch but @nicowilliams and I are looking at it.

@SteveSyfuhs wrote:

> This is not our problem. Our KDC, our client, and a handful of other well-known implementations have no issue with this change.

I agree this is not a security issue.  Its an interop issue with bad outcomes for end users.   Apple KDCs will have the same issues.  I and other Heimdal developers have worked with Microsoft on Kerberos since 1997 in the best interests of end users.   I would hope that Microsoft would take that into account when making changes to their implementation.

FYI, RFC4120 suggests that if the desired request is for the maximum ticket lifetime the request should request a till time of UNIX Epoch 19700101000000Z.  This value is called out in RFC 4120 for that purpose.

**Comment by jaltman** [comment]
@greghudson wrote:
> I'm not sure if "a handful of other well-known implementations" refers to MIT krb5 here. MIT krb5 can represent timestamps up to 21060207062815Z, as of release 1.16 (prior to that it is limited to y2038).

@greghudson, @nicowilliams and I have looked at converting krb5_timestamp from signed 32-bit integer to unsigned 32-bit integer in order to extend the maximum lifetime to 2016.  However, I am concerned about the change because krb5_timestamp is embedded in public data structures and as a public parameter type in several public and even some so called private but exported functions.   When MIT made the switch in 1.16 did the team consider the impact on applications that have expectations that the krb5_timestamp is a signed 32-bit integer?   And what might happen if the values wrap begin wrapping negative?

**Comment by SteveSyfuhs** [comment]
> I would hope that Microsoft would take that into account when making changes to their implementation.

To clarify my point, we did our due-diligence against a wide swath of test cases that didn't exhibit any problems. So far this particular branch of Heimdal is the only thing exhibiting issues. We are not inclined to revert this change without a good justification (meaning real customer impact), but we are willing to work with this team to find a reasonable upper limit that will go into the next major release (any backport of that change would still require the same justification).

**Comment by jborean93** [comment]
Just as an FYI it seems like a Samba KDC uses an affected branch of Heimdal. I'm unsure if they've patched it themselves but I've seen a few reports where Kerberos auth no longer works on the latest Win 11 versions and there "workaround" is to unfortunately enable DES encryption.

**Comment by lhoward** [comment]
> Just as an FYI it seems like a Samba KDC uses an affected branch of Heimdal. I'm unsure if they've patched it themselves but I've seen a few reports where Kerberos auth no longer works on the latest Win 11 versions and there "workaround" is to unfortunately enable DES encryption.

That's a very unfortunate workaround and I'm puzzled how that would fix things (although, from memory, the DES flag enables a bunch of other behaviours). It looks like the re-encoding issue was fixed in ebfd48e40a1b61bf5a6b8d00fe5c581e24652b6e, I'm not sure if there are any other issues, but back-porting that fix would be a start. If Samba has that commit then, clearly, there are other issues to resolve.

**Comment by jborean93** [comment]
> That's a very unfortunate workaround

Very much so, I've neither tried to replicate it or confirmed it actually works but here are some random Reddit posts where people state it was their workaround

* https://www.reddit.com/r/windowsinsiders/comments/t1f7hu/cannot_connect_to_samba_ad_dc_on_windows_11_dev/
* https://www.reddit.com/r/synology/comments/xk2a7q/psa_windows_11_22h2_incompatible_with_synology/

I just assumed DES was so old it didn't do the same sort of checks on the messages but that's very much a stab in the dark. I just thought it a good idea to let the more knowledgeable people know some of the real life scenarios where this might be a problem.

**Comment by lhoward** [comment]
Yeah, it's quite possible DES didn't use keyed checksums, perhaps. But... we definitely need to get onto this before insecure workarounds start floating around on Reddit.

**Comment by jaltman** [comment]
Backporting https://github.com/heimdal/heimdal/commit/ebfd48e40a1b61bf5a6b8d00fe5c581e24652b6e to heimdal-7-1-branch is the next step.  There is not a clean cherry-pick but I hope to be able to pull the required bits out.   Its getting late on the East Coast so I will work on that tomorrow if no one else does first.

**Comment by greghudson** [comment]
> However, I am concerned about the change because krb5_timestamp is embedded in public data structures and as a public parameter type in several public and even some so called private but exported functions. When MIT made the switch in 1.16 did the team consider the impact on applications that have expectations that the krb5_timestamp is a signed 32-bit integer? And what might happen if the values wrap begin wrapping negative?

I did consider it.  Some libkrb5 applications may still have y2038 issues and require their own fixes.  The other approaches I considered didn't seem likely to work out better, though.

**Comment by akarl10** [comment]
> Yeah, it's quite possible DES didn't use keyed checksums, perhaps. But... we definitely need to get onto this before insecure workarounds start floating around on Reddit.

I tried this with ubuntu 22.04 and this workaround is worse than you think: des is the same as disabling kerberos altogether. Easily observable by checking klist on the windows 11 side (that is empty). Only ntlm will be in used. I think this only works because not checking anything in the group policy dialogue is not possible/enables everything so you need to check something that surely gets rejected in the first place (arc4 would still work, but then you trigger the issue mentioned here)

**Comment by abartlet** [comment]
Samba will track this as https://bugzilla.samba.org/show_bug.cgi?id=15197

I've not dug into this yet but all reports are that the maintained versions of Samba (4.17 and 4.16) operate correctly.

While Samba 4.15 is in security-fixes-only mode, hopefully the suggested backport approach works as desired, as then we can provide that for users still using the older versions.

**Comment by Neustradamus** [comment]
To follow this ticket

**Comment by abartlet** [comment]
I can confirm the issue reproduces on Windows_InsiderPreview_Server_vNext_en-us_25217 which is easier to put in a VM than Windows 11 22H2 and easier to download (available for public download as a testing resource).

I'm now going to attempt the backport suggested by @jaltman at https://github.com/heimdal/heimdal/issues/1011#issuecomment-1258921169 for Samba (and will of course provide the results back to Heimdal).

**Comment by abartlet** [comment]
I've manually confirmed that Samba 4.16 and 4.17 are not impacted, as they use Heimdal a 0.8pre snapshot.  Samba 4.15 and earlier is impacted, and a Samba patch is now on our bug https://bugzilla.samba.org/show_bug.cgi?id=15197

**Comment by iio7** [comment]
I don't know if this is related, but I have just setup an OpenBSD 7.2 machine running Heimdal 7.7.0 as a Kerberos server. I then have a NFS Linux server on another machine and a FreeBSD NFS client on another machine.

When I try to mount the NFS share on the Linux machine from the FreeBSD client, I get the following in the log on the OpenBSD Heimdal server:

```
Oct 29 00:16:54 foo kdc[55215]: Failed to verify AP-REQ: Decrypt integrity check failed for checksum type hmac-sha1-96-aes256, key type aes256-cts-hmac-sha1-96
Oct 29 00:16:54 foo kdc[55215]: Failed parsing TGS-REQ from IPv4:192.168.1.4
Oct 29 00:16:54 foo kdc[55215]: tgs-req: sending error: -1765328353 to client
Oct 29 00:16:54 foo kdc[55215]: sending 81 bytes to IPv4:192.168.1.4
```

**Comment by SteveSyfuhs** [comment]
Howdy folks. We have modified the upper in Windows to be `2100`. The value `0` evidently has uglier compatibility issues... We are not willing to return the value to 2038.

The change is making its way through our branches. It will arrive in Insiders dev builds within a couple weeks at the earliest. Practically speaking it'll probably take longer. Will keep folks apprised of the timeline.

Given that this is only affecting W11 22H2, we are pushing to fix it in-market. Before we commit to doing that, we'd like verification from this team that the change does in fact fix the issue. Specifically, we're asking for the following:

1. I will indicate on this bug once the Insiders build with the fix becomes available (ETA 2-4 weeks)
2. Someone here installs that build and attempts a full AS/TGS/AP against the affected version of Heimdal
3. Reports back OK/BAD

We try to avoid releasing non-security fixes in the Nov/Dec timeframe because many organizations are in lockdown. Assuming all the stars align, we'd likely see this change hit in-market sometime between Jan/Mar 2023. That is not a guarantee -- just being transparent.

**Comment by greghudson** [comment]
> The value 0 evidently has uglier compatibility issues...

For my own reasons I would be interesting to hear what those are.

**Comment by jaltman** [comment]
@SteveSyfuhs wrote

> Howdy folks. We have modified the upper in Windows to be `2100`. The value `0` evidently has uglier compatibility issues... We are not willing to return the value to 2038.
>     3. Reports back OK/BAD

Thank you for making this change but a year of 2100 will still cause a failure.   As mentioned earlier in the discussion:

> The Heimdal ASN.1 engine does not support timestamps larger than 0x7FFFFFFF seconds from UNIX Epoch.

  2100 x 365.25 x 24 x 60 x 60 = 66270960000 = 0xF:6E0E9980

which exceeds the supported limit.   By itself, that change will not prevent the failure.
Earlier today I merged https://github.com/heimdal/heimdal/commit/4cd3926e14eca91ca10a9ec87200c8bd717b66e4 which permits the request time to be capped without causing an authenticator verification failure.

This will be included in the Heimdal 7.9 release.

@SteveSyfuhs  wrote:
> The value 0 evidently has uglier compatibility issues...

I too would like to hear what they are.   It would be worth adding an errata to the RFC if zero does not work.

**Comment by sbraeuning** [comment]
Possible to backport this to Heimdal 7.7 as well? Debian Bullseye stable is currently using this version in combination with samba 4.13 ...

**Comment by SteveSyfuhs** [comment]
Just as an FYI the change to 2100 goes live in [Insiders this week](https://blogs.windows.com/windows-insider/). Waiting to get the official build number that it'll be in.

> The Heimdal ASN.1 engine does not support timestamps larger than 0x7FFFFFFF seconds from UNIX Epoch.
> 2100 x 365.25 x 24 x 60 x 60 = 66270960000 = 0xF:6E0E9980
> which exceeds the supported limit. By itself, that change will not prevent the failure.

Thanks for the clarification @jaltman. Can you describe the failure behavior specifically (e.g. what error path it takes)? My understanding is that the number would overflow, producing the incorrect logical value, but the checksum wouldn't fail, and the system would proceed as normal. Is that an incorrect understanding?

Our concern here is unblocking customers from logging into Heimdal-protected resources. If folks can pull in the latest patch on their network devices then I'm less concerned about this exact scenario and am more concerned about longer term interoperability post 2038. I suspect there will be all sorts of cases where that isn't feasible though.

> I too would like to hear what they are. It would be worth adding an errata to the RFC if zero does not work.

This was more speculation and open source code analysis than actual testing. We've had it hardcoded as a date in the future for so long it's almost certainly going to cause problems.

**Comment by SteveSyfuhs** [comment]
Howdy folks. We have an updated build available now: https://blogs.windows.com/windows-insider/2022/11/18/announcing-windows-11-insider-preview-build-25247/

**Comment by vmsman** [comment]
Any idea when a patch to Windows 11 will occur?

**Comment by jaltman** [comment]
> Possible to backport this to Heimdal 7.7 as well? Debian Bullseye stable is currently using this version in combination with samba 4.13 ...

The change has already been cherry-picked to the Heimdal 7.x release branch and will be included in Heimdal 7.9.

**Comment by NuAngel** [comment]
Will the fixes for this be backported to Ubuntu 21.04.0x LTS branches in the near future? Many companies use the LTS server, and with the 22H2 feature updates being rolled out to more and more computers, or even buying new hardware with it pre-installed, companies like mine are finding ourselves in a bad situation.

**Comment by NuAngel** [comment]
@SteveSyfuhs if the fix is available in Inisder Preview build 25247, can we receive a Hotfix for existing 22H2 machines? How long until we should expect this update to become available to non Insider Preview computers???

**Comment by panlinux** [comment]
> Will the fixes for this be backported to Ubuntu 21.04.0x LTS branches in the near future? Many companies use the LTS server, and with the 22H2 feature updates being rolled out to more and more computers, or even buying new hardware with it pre-installed, companies like mine are finding ourselves in a bad situation.

This is the Ubuntu tracking bug for this: https://bugs.launchpad.net/ubuntu/+source/samba/+bug/1993934

**Comment by vmsman** [comment]
I am really amazed that this is not a hotfix by now.  I have paused a rollout at a firm that I support for weeks now.

**Comment by abartlet** [comment]
Remember that (if you consume Heimdal via Samba) the two most recent
and the only two maintained (as compared to security-fixes-only) Samba
versions don't have any problem with the strange 9999 date.

**Comment by NuAngel** [comment]
Thank you @panlinux! Looking forward to this Samba fix to be backported to Ubuntu VERY soon. This DOES make me wish I was using Debian and not Ubuntu, @abartlet! lol Too bad I was just hired to support what they had - rebuilding from scratch is an option that might be _years_ away!

**Comment by panlinux** [comment]
> Thank you @panlinux! Looking forward to this Samba fix to be backported to Ubuntu VERY soon. This DOES make me wish I was using Debian and not Ubuntu, @abartlet! lol Too bad I was just hired to support what they had - rebuilding from scratch is an option that might be _years_ away!

You are welcomed to test the packages, they are in proposed, for focal (20.04) and jammy (22.04). Bionic (18.04) was halted because of a quick question, but which was answered already, so it should be accepted soon.

**Comment by SteveSyfuhs** [comment]
Howdy folks. Just wanting to let you know we've tentatively slotted the Windows change for the last week of February. For folks that don't take that particular update, it'll show up on Patch Tuesday in March or April. Usual caveat that there's no guarantee it will be in any particular update cycle until that cycle actually ships.

**Comment by sbraeuning** [comment]
> Howdy folks. Just wanting to let you know we've tentatively slotted the Windows change for the last week of February. For folks that don't take that particular update, it'll show up on Patch Tuesday in March or April. Usual caveat that there's no guarantee it will be in any particular update cycle until that cycle actually ships.

@SteveSyfuhs 
Will it be possible to provide the update number (KB...) within this post, when the update has been launched?

**Comment by SteveSyfuhs** [comment]
> Will it be possible to provide the update number (KB...) within this post, when the update has been launched?

I will post it if I remember. That said, all updates are cumulative. If you don't take the 2D optional update, it'll be slipstreamed into the 3B or 3C update, and that will be slipstreamed into the 4B update. Get current -- stay current mantra.

**Comment by sbraeuning** [comment]
Fix is now part of regular quality update "2023-03 Cumulative Update for Windows 11 Version 22H2 for x64-based Systems (KB5023706)" which is available since 14th march 2023.

**Comment by ThatWileyGuy** [comment]
Worth a mention that this change made it into Server 2025 as well and people are starting to encounter it in KDC messages.

**Comment by greghudson** [comment]
> These are the same things.

If I understand the situation correctly, I would argue that the AS-REP behavior is significantly worse because (1) it affects a larger class of endpoints, and (2) I cannot currently imagine any justification for it (whereas the AS-REQ `till` behavior is perhaps justified by claimed compatibility issues with using the RFC4120-indicated `19700101000000Z`).

For AS-REP I believe the far-future timestamp is being used in a `LastReq` sequence element with type 6 or 7 to indicate the password and/or account expiration time.  If there is no expiration time to report, why include the sequence element at all?

**Comment by elric1** [comment]
The particular bug raised is fix, but here are still outstanding 2038 issues. #1231 will track those and I shall close this.

**Comment by Neustradamus** [comment]
Linked to:
- https://github.com/heimdal/heimdal/issues/533
- https://github.com/heimdal/heimdal/pull/726
- https://github.com/heimdal/heimdal/pull/743
- https://github.com/heimdal/heimdal/issues/1011
- https://github.com/heimdal/heimdal/issues/1231
- https://github.com/heimdal/heimdal/issues/1385
